### ISSUE TITLE
Error "Please add a @NgModule annotation" in Angular4

### ISSUE THREAD

**Issue by davidanaya** [angular/angular#15763]
**I'm submitting a ...**  (check one with "x")
```
[X] bug report
```

**Current behavior**
I have an Angular application which imports a module from another repository.

So my app.module.ts
```
@NgModule({
  imports: [
    ...
    MyCommonModule,
  ],
  declarations: [...],
  providers: [...],
  bootstrap: [AppComponent]
})
export class AppModule { }
```
and then in my node_modules I have my common repository (installed with yarn/npm in package.json) with the file file my-common-module.d.ts
```
@NgModule({
  imports: [...],
  declarations: [...],
  providers: [...],
  exports: [...]
})
export class MyCommonModule {}
```

This works fine with Angular 2.
Now, I delete node_modules and change the version of Angular to 4.0 in the package.json of my app and then yarn again.
I build the app with webpack, no errors, and when I run it I get the following error in the browser
```
Uncaught Error: Unexpected value 'MyCommonModule' imported by the module 'AppModule'. Please add a @NgModule annotation.
```

**Expected behavior**
I'd expect this to work the same with Angular4 or Angular2.

My question is, do I need to also upgrade the common module to Angular4 for this to work? I thought that this would already be transpiled to js when I import it in my main app module, maybe I'm wrong.

**Please tell us about your environment:**
I tested this with webpack-dev-server

* **Angular version:** 2.0.X
With 2.0.0 it's working, with 4.0.0 it's not.

* **Browser:**
I tested this in Chrome and FF.
 
* **Language:**
I'm using typescript 2.2.2

* **Node (for AoT issues):** `node --version` =

**Comment by DzmitryShylovich** [comment]
Please add a small reproduction.

**Comment by davidanaya** [comment]
I can´t create a plunkr with this setup nor upload this to my git account as it's not open source code. I could try to create another 2 simple repos to reproduce this same scenario, but I'd need to confirm first that I can import a repo generated with Angular 2 from another project in Angular 4. That should not be an issue, right?
Thanks!

**Comment by DzmitryShylovich** [comment]
> but I'd need to confirm first that I can import a repo generated with Angular 2 from another project in Angular 4

yes, it should work without any issue.

is `MyCommonModule` your own or 3rd party module?

**Comment by davidanaya** [comment]
It's my own module. It contains some common components used by different projects.
I wanted to upgrade one of those projects to Angular 4 but without touching the commons one.

**Comment by kukjevov** [comment]
What are your dependencies on your `MyCommonModule`? How did you specify dependency on angular? It is among dependencies or peerDependecies? Anyway if you have something like this `"@angular/core": "^2.4.3"` in your "library" and `"@angular/core": "^4.0.1"` in your application it wont work. NPM creates two versions of angular in your node_modules and if `NgModules` is defined in 2 files they are not same. 

You have to set version in your "library" to `"@angular/core": ">=2.4.3"` or to `"@angular/core": "^4.0.0"`. You should check subdirectory of your lib in node_modules if there is directory named node_modules and if there is `@angular/core`.

If it is there, that is the problem. I hope it helps, if not i can explain better :).

**Comment by Toxicable** [comment]
It's possible that the way you're bundling your External Module is incorrect. It sounds like to me that your External Module is targeting v2 therefore v2 @NgModule does not equal v4 @NgModule hence the error.
Either way, this does not sound like a bug with the framework, it sounds like a support request. This issue tracker is not suitable for Support Requests.

**Comment by davidanaya** [comment]
Thanks @kukjevov so it really is a problem if my library uses Angular 2 and my app uses Angular 4.
The thing is my library in Angular 2 is used by other applications which are in Angular 2 and I can´t migrate them at the moment, so it has to stay in Angular 2.

If I use ```>=2.4.3``` in my library does it mean that it will use Angular 4 when I build my app and Angular 2.4.3 when I use it with my other apps?

@Toxicable thanks for your input, I think that's exactly what happens I think.

**Comment by robwormald** [comment]
Any component written for 2.x should work for 4.x. 

It's important that you do not generate the factories in your library code. Your library code should only emit the metadata.json files. 

Each application should then import NgModules from the shared library code, and compile each application as a unit. 

This is baked into the entire design of angular's compiler - backwards compatibility. This works because the metadata.json is backwards compatible, and ngFactory code can be generated for each version when an app is compiled.

**Comment by davidanaya** [comment]
Thanks @robwormald could you please elaborate or give me a hint on where to look for some more info about it? My common module inside node_modules has a my-common-module.d.ts with the exports, including ```export declare class MyCommonModule {}``` and then the transpiled contents of the module in js, but I never heard of metadata.json files.

**Comment by kukjevov** [comment]
hi @davidanaya  well if you use `>=2.4.3` your library might compile with latest angular, but if you use this library that has dependency spefified this way it should work with any version of angular higher than 2.4.3. That means if your application has dependency `"@angular/core": "2.4.5"` your library should not create its own node_modules with different angular version since it is compatible with version specified in application package.json. Also if your application uses version `4.0.1` it is still in range of `>=2.4.3` and it should be fine.

I am not completely sure, but i think this should work as far as your code will be backward compatible. It does not matter if your library is compiled with 2.4.3, or 4.0.1 if code compiles and works with both of them.

**Comment by davidanaya** [comment]
Thanks @kukjevov I'll give it a try.

**Comment by davidanaya** [comment]
Well, that worked!
Thanks a lot, @kukjevov 

So, just to summarize in case someone else needs it.

In my common library I used this in my package.json

```
 "dependencies": {
    "@angular/core": ">=2.0.0",
    "@angular/common": ">=2.0.0",
    "@angular/router": ">=3.0.0",
    "@angular/http": ">=2.0.0",
    "@angular/platform-browser": ">=2.0.0",
```

I did this in a different branch, ```feature/angular4``` and run ```yarn install```
When I checked my yarn.lock I saw this was inside:
```
"@angular/common@>2.0.0":
  version "4.0.1"
  resolved "https://registry.yarnpkg.com/@angular/common/-/common-4.0.1.tgz#df488eada842b2d841ded750712292b18387b5b0"
```

Looks promising; if I'm not wrong it took the last available version for @angular/common, which is 4.0.1

Now I push the branch to my github repo and change to my app project.

I run ```yarn upgrade my_guthub_repo/common#feature/angular4```
I check my yarn.lock and I have this
```
"@angular/common@4.0.0", "@angular/common@>2.0.0":
  version "4.0.0"
  resolved "https://registry.npmjs.org/@angular/common/-/common-4.0.0.tgz#ca18983222fdab4ecaa7a8b99eda6ff661e6dc92"
```

I run it, and it works 👍

**Comment by vdzundza** [comment]
Hi guys, I've the same issue. My angular version is 4.0.1
There are repo with reproducible error https://github.com/vdzundza/travis-ng2/

**Comment by IevhenIkonnykov** [comment]
Hello,
It seems that I have a similar problem, but it looks differently.
I had a project on Angular 2.4.3 that worked fine and then I moved it to Angular 4.0.1. So now I have a problem:

**compiler.es5.js:14577** Uncaught Error: Unexpected value 'LoadingComponent' declared by the module 'AppModule'. Please add a @Pipe/@Directive/@Component annotation.
    at syntaxError (http://localhost:4200/vendor.bundle.js:53165:34) [<root>]
    at http://localhost:4200/vendor.bundle.js:65680:40 [<root>]
    at Array.forEach (native) [<root>]
    at CompileMetadataResolver.getNgModuleMetadata (http://localhost:4200/vendor.bundle.js:65662:54) [<root>]
    at JitCompiler._loadModules (http://localhost:4200/vendor.bundle.js:76740:64) [<root>]
    at JitCompiler._compileModuleAndComponents (http://localhost:4200/vendor.bundle.js:76699:52) [<root>]
    at JitCompiler.compileModuleAsync (http://localhost:4200/vendor.bundle.js:76661:21) [<root>]
    at PlatformRef_._bootstrapModuleWithZone (http://localhost:4200/vendor.bundle.js:5145:25) [<root>]
    at PlatformRef_.bootstrapModule (http://localhost:4200/vendor.bundle.js:5131:21) [<root>]
    at Object.245 (http://localhost:4200/main.bundle.js:16726:124) [<root>]
    at __webpack_require__ (http://localhost:4200/inline.bundle.js:53:30) [<root>]
    at Object.778 (http://localhost:4200/main.bundle.js:45437:18) [<root>]
    at __webpack_require__ (http://localhost:4200/inline.bundle.js:53:30) [<root>]
    at webpackJsonpCallback (http://localhost:4200/inline.bundle.js:24:23) [<root>]

But LoadingComponent is correct:

import {Component, Input} from '@angular/core';
@Component({
  selector: 'cdp-loading',
  templateUrl: './loading.component.html'
})
export class LoadingComponent {
  @Input() public isBusy: boolean;
}

When I remove this component it complains on another and so on.
My package.json:
{
  "name": "web-services-ui",
  "version": "0.1.0",
  "description": "Angular2 SPA",
  "license": "ISC",
  "author": "Author",
  "scripts": {
    "ng": "ng",
    "start": "ng serve --proxy-config proxy.config.json",
    "build": "ng build",
    "lint": "ng lint",
    "test": "ng test",
    "e2e": "protractor"
  },
  "private": true,
  "dependencies": {
    "@angular/common": "^4.0.0",
    "@angular/compiler": "^4.0.0",
    "@angular/core": "^4.0.0",
    "@angular/forms": "^4.0.0",
    "@angular/http": "^4.0.0",
    "@angular/platform-browser": "^4.0.0",
    "@angular/platform-browser-dynamic": "^4.0.0",
    "@angular/platform-server": "^4.0.0",
    "@angular/router": "^4.0.0",
    "@angular/animations": "^4.0.0",
    "core-js": "^2.4.1",
    "rxjs": "^5.1.0",
    "zone.js": "^0.8.4",
    "primeng": "^4.0.0-rc.2",
    "marked": "0.3.6",
    "prismjs": "^1.6.0"
  },
  "devDependencies": {
    "@angular/cli": "1.0.0",
    "@angular/compiler-cli": "^4.0.0",
    "@types/jasmine": "2.5.38",
    "@types/node": "~6.0.60",
    "codelyzer": "~2.0.0",
    "jasmine-core": "~2.5.2",
    "jasmine-spec-reporter": "~3.2.0",
    "karma": "~1.4.1",
    "karma-chrome-launcher": "~2.0.0",
    "karma-cli": "~1.0.1",
    "karma-jasmine": "~1.1.0",
    "karma-jasmine-html-reporter": "^0.2.2",
    "karma-coverage-istanbul-reporter": "^0.2.0",
    "protractor": "~5.1.0",
    "ts-node": "~2.0.0",
    "tslint": "~4.5.0",
    "typescript": "~2.2.0"
  }
}

My AppModule:

@NgModule({
  declarations: [
    AppComponent,
    OpenDirective,
    HeaderComponent,
    NavigationSiteDirectoryComponent,
    NavigationComponent,
    FooterComponent,
    ContentComponent,
    HomeComponent,
    LoadingComponent,
    OpenEngineeringModelModalComponent,
    LogoutComponent,
    ModalContainerComponent,
    ModelNavigationContainerComponent,
    NavigationModelComponent,
    ControlSideBarComponent
  ],
  imports: [
    BrowserModule,
    HttpModule,
    APP_ROUTER_PROVIDERS,
    FormsModule,
    ReactiveFormsModule,
    CommonModule,
    CDP4SharedModule,
    RequirementModule,
    SiteDirectoryModule,
    EngineeringModelModule,
    RIDModule,
    EngineeringModelDataNoteModule,
    FileStoreModule
  ],
  providers: [
    {provide: APP_BASE_HREF, useValue: '/'},
    BusyIndicatorService,
    ThingUpdateService,
    WebServiceClient,
    CdpServicesDal,
    ModelNavigationService,
    UserSessionService,
    SiteDirectoryResolver,
    EngineeringModelResolver,
    DialogService,
    ValidationService,
    TreeTableService,
    RequirementTreeTableService,
    SharedTreeTableService,
    FileStoreTreeTableService,
    PersonTreeTableService,
    ConfigService,
    {provide: APP_INITIALIZER, useFactory: initConfig, deps: [ConfigService], multi: true }
  ],
  bootstrap: [AppComponent]
})

export class AppModule {
}

Any ideas why it might happen and how to solve it?
Thanks.

**Comment by IevhenIkonnykov** [comment]
Just to add - when I delete a few components I finish with a problem described [here](https://github.com/angular/angular-cli/issues/5007).
In both cases I use Angular cli.

**Comment by helzgate** [comment]
Are you guys saying that any 3rd party library that has "@angular/core":"^2.0.0" (as an example) in it's package.json won't work when imported as a module?  This is a nightmare, none of my 3rd party libraries work if this is the case.

**Comment by helzgate** [comment]
Never mind. My issue was related to the DLLReferencePlugin in Webpack.  I had to clean my dll directory and rebuild them.

**Comment by afisher88** [comment]
You can also get this error if you declare a component in the imports array.

**Comment by Lavaei** [comment]
Maybe you have "node_modules" directory in your project!

**Comment by FriOne** [comment]
Hello! I have the same problem, I forked a library to make it work with angular cli aot, but no luck.
Could anyone look at [my fork](https://github.com/FriOne/ng2-flatpickr)?
An error is 

> ERROR in Unexpected value 'Ng2FlatpickrModule in .../node_modules/ng2-flatpickr/lib/ng2-flatpickr.module.js' imported by the module 'AppModule in .../src/app/app.module.t
> s'. Please add a @NgModule annotation.

**Comment by prateekvarma** [comment]
I'm facing a similar issue. However, i get this error when i try to register a service on my Angular4 app. The error i see strangely is for another component (which works fine when the service is removed) :/

**Comment by dougludlow** [comment]
I ran into this using angular-cli and after quite a bit of digging found that it had to do with the generated `polyfills.ts`. In it, it imports the `core-js` versions of reflect:

```typescript
import 'core-js/es6/reflect';
import 'core-js/es7/reflect';
```

I commented those out and imported `reflect-metadata`:

```typescript
import 'reflect-metadata';
```

Things started working just fine after. Not exactly sure why this works... my understanding was that `core-js` simply included the `reflect-metadata` polyfill...

**Comment by AckerApple** [comment]
I was searching all over for tips on this error. Turns out I forgot to add "typings":"index.d.ts" to my package.json file

**Comment by irontoby** [comment]
I was having this error which turned out to be as @Lavaei suggested. I was doing local development on my shared module in a different project, and used `yarn link` to use the local copy. I had run `yarn install` in the linked copy, to make development easier (intellisense in VS Code, etc.)

As soon as I deleted `node_modules` from the local shared project, this error went away. I'm also using Angular CLI.

Is it possible to tell Angular CLI to ignore these `node_modules` in my local copy of the shared repo? It would be nice to be able to both serve this shared repo directly (I have a "shell app" that displays some of the common components) as well as use it from a different repo with `yarn link`.

**Comment by etwillbefine** [comment]
Can confirm @irontoby 
I had the same issue and after removing `node_modules` and reinstall the error was gone. 
Using angular-cli as well. It happen for me after upgrading from 4.1.0 to 4.2.0-rc.0

**Comment by CrazyIrish** [comment]
@dougludlow thanks, this was my issue as well.

**Comment by btgoodwin** [comment]
@etwillbefine Checking in to confirm I have the same problem as described, however if I reinstall (`npm install`) for the library, the error returns.  

I'm in a development environment where I use `npm link` in my library's `dist` path (i.e., where build artifacts are placed), and then at the application run `npm link my_library`.   Previously  this worked fine, but my ModuleWithProviders configuration broke from angular-cli 1.0.1 to 1.1.3.  So I fixed that error so that 1.1.3 seems to be fine, but now I get this error.  

For what it's worth, the linked library's `dist` directory does not have a `node_modules` path; it's outside the tree.  It's unclear to me how to work around this new problem just to restore the simplicity of using `npm link`.

**Comment by stevencohn** [comment]
I can confirm the findings: deleting node_modules from the shared library works. What's strange is that I did NOT change the version of angular or angular-cli but rather upgraded my Node.js installation. So there's some new relationship with probably tsconfig moduleResolution and that combinations of newer Node + angular-cli installs lock the shared library node_modules directory and makes the run time confused. For now, I can automate deleting or hiding those local node_modules directories but it's a real PITA.

**Comment by stevencohn** [comment]
So it seems the difference was in how npm install is implemented and changed from npm v4 to v5 and how Node discovers relevant node_modules folders. Previously, npm install would copy a local shared libraries dist folder into the consuming project's node_modules. Now it instead creates a symlink inside the consuming project's node_modules.

So previously with a copied shared library dist folder, Node would run and crawl up the tree looking for node_modules, which it found immediately because that subfolder was already the consuming project's node_modules. Now, when Node runs and crawls up the tree from the symlink, it discovers the shared library's node_module folder from within that shared library project folder. (It's being a little too smart knowing where the symlink points rather than where the symlink is defined.)

My solution was to change where my shared libraries generate their dist folders. They now generate the dist folder outside of their own project folder, beyond the reach of their own node_modules. I now reference that external dist folder when using the npm link command and npm install references the file:// target of the link command. Now when Node run the Angular application, it won't get confused by finding these irrelevant node_modules folders.

To be fair, I like that they use symlinks instead of copying into node_modules because this allows quicker local development of a shared library without having to reinstall it every time you want to test a change. So I don't think it's a problem; we just need to understand how it works and work with it!

**Comment by ruslan-simonenko** [comment]
I've had an issue similar to @stevencohn 's one, and I was able to fix it by installing my dependencies through [yarn](https://yarnpkg.com/lang/en/docs/migrating-from-npm/) instead of `npm`. I've lost `npm link` functionality, but the code started working with multiple locally-built libraries.

**Comment by hrushikesh07** [comment]
Unexpected value 'DataTableModule' imported by the module 'AppModule'. Please add a @NgModule annotation.

**Comment by kmolerov** [comment]
I can confirm that mentioned solutions from above are working. 
Problem is caused from using symlink after running  npm link [your-lib] if there is populated node_modules folder in library project (and dist is a subdirectory of it).

Solution1: Remove symlink and then copy dist folder into node_modules.
Solution2: Delete node_modules from library project.
**Solution3**: Build library into dist folder that is not inside library project.

**Comment by omouse** [comment]
None of the above solutions worked for me, I still get this error:

    Error: Unexpected value 't' imported by the module 't'. Please add a @NgModule annotation.  build.js:1:44466

**Comment by zhanggongze** [comment]
Delete node_modules page and re Download

**Comment by dohpaz42** [comment]
Slightly the same, but mostly different:

To give my two cents, I am using Angular 4.3.5, and I have a component in my main app module that I wanted to use in a local (non-npm) module. I had mistakenly added it to my local module's ngModule's import definition, which gave me this error:

```javascript
Uncaught Error: Unexpected directive 'SharedComponent' imported by the module 'LocalModule'. Please add a @NgModule annotation.
```

Once I refactored my `@ngModule` definition, so that my `LocalModule` was in both the declarations AND the exports, it worked like a charm.

HTH

**Comment by omouse** [comment]
@dohpaz42 could you show the code for that?

**Comment by dohpaz42** [comment]
@omouse, sure:

### Original AppModule definition:
```javascript
import { SharedComponent } from './_shared/shared.component';

@NgModule({
    declarations: [SharedComponent]
})
export class AppModule { }
```

### Updated AppModule
```javascript
import { SharedComponent } from './_shared/shared.component';

@NgModule({
    exports: [SharedComponent]
})
export class AppModule { }
```
### LocalModule
```javascript
import { SharedComponent } from 'app/_shared/shared.component';
@NgModule({
    declarations: [SharedComponent]
    exports: [SharedComponent]
})
export class LocalModule { }
```

**Comment by playground** [comment]
@robwormald I followed your example in creating a shared lib and it has been working in my project.  When I tried to create the same shared lib using https://github.com/filipesilva/angular-quickstart-lib/ and import it into demo app, I'm getting this error "Please add a @NgModule annotation".  Can you provide some insight?  

I have filed an issue here
https://github.com/filipesilva/angular-quickstart-lib/issues/55

**Comment by playground** [comment]
@dohpaz42, I'm still getting the same error.  My shared module is just a service with no ui component.

Shared Service Module
```
import { ShareService } from './service/share.service';
@NgModule({
  declarations: [],
  providers: [
    ShareService
  ],
  imports: [],
  exports: []
})
export class ShareServiceModule { }
```
Local Module
```
import { ShareServiceModule} from 'share-service';
@NgModule({
  imports: [ShareServiceModule]
})
export class AppModule { }
```

**Comment by dohpaz42** [comment]
@playground not sure if it works the same as a Component, but you a) didn't export the service from AppModule, and 2) you didn't add your service in the declarations and exports of your other module. If that doesn't work, you may try opening a new ticket so the Angular devs can look into it (this issue is closed and probably ignored by them, my guess).

**Comment by ahermle93** [comment]
I'm also facing this problem in my own 3rd-party modules.

A little workaround for me is to add manually the @Component and @Input() annotations in the d.ts files. These annotations go missing when the d.ts files are generated from the .ts source files.

But this ain't a suitable solution so is there any better approach to solve it?

**Comment by jsgoupil** [comment]
I got this error if I import "reflect-metadata"; **again**
this will clear the Metadata cache and Angular will not be able to find the ngModule decorator appropriately.

I removed my `import 'reflect-metadata';` and used `window['Reflect'].getMetadata()` in my script.

**Comment by staskolukasz** [comment]
This kind of error can also occur during the tests, when in your imports you declare "SomethingComponent" instead of "SomethingModule".

**Comment by mohanasundaramrebel** [comment]
i am getting this  error how do i fix this, i am new to angular 

Error: Unexpected value 'Http' imported by the module 'AppModule'. Please add a @NgModule annotation.  vendor.js:78125:34
	syntaxError http://localhost:8100/build/vendor.js:78125:34
	CompileMetadataResolver.prototype.getNgModuleMetadata/< http://localhost:8100/build/vendor.js:91973:44
	forEach self-hosted:267:13
	CompileMetadataResolver.prototype.getNgModuleMetadata http://localhost:8100/build/vendor.js:91956:13
	JitCompiler.prototype._loadModules http://localhost:8100/build/vendor.js:103389:47
	JitCompiler.prototype._compileModuleAndComponents http://localhost:8100/build/vendor.js:103362:31
	JitCompiler.prototype.compileModuleAsync http://localhost:8100/build/vendor.js:103291:32
	PlatformRef_.prototype._bootstrapModuleWithZone http://localhost:8100/build/vendor.js:4893:16
	PlatformRef_.prototype.bootstrapModule http://localhost:8100/build/vendor.js:4879:16
	[196] http://localhost:8100/build/main.js:191:1
	__webpack_require__ http://localhost:8100/build/vendor.js:55:12
	webpackJsonpCallback http://localhost:8100/build/vendor.js:26:23
	<anonymous> http://localhost:8100/build/main.js:1:1

**Comment by IevhenIkonnykov** [comment]
@mohanasundaramrebel It would be great to have more explanation. Is it yours Http module or Angular's? First check where you imported it and whether it is a correct name. Does it work in aot mode?

**Comment by Basel78** [comment]
I have library built in angluar4 and wanted to consume it angular-cli project
when i reference the my lib in the project app.module.ts or app.routing.ts error throwing in runtime  and sometime when run ng build
```
Uncaught Error: Unexpected value 'AppRoutingModule' imported by the module 'AppModule'. Please add a @NgModule annotation.
```
after some research i find out how to do that using Yomen with "generator-angular2-library".
and from the Sample they have i find a little difference on how the NgModule is generated  where they are using the following code to define and expose the components
```
@NgModule({
  imports: [
    CommonModule
  ],
  declarations: [
    SampleComponent,
    SampleDirective,
    SamplePipe
  ],
  exports: [
    SampleComponent,
    SampleDirective,
    SamplePipe
  ]
})
export class SampleModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: SampleModule,
      providers: [SampleService]
    };
  }
}
```
so I changed my code to comply with same format they have used, but the error but that didn't help resolving the error,
I conduct debugging from from browser to the angular-cli code and find my library code in vendor.bundle.js file and the components were initiated properly, but when the App run bootstraping function some code fails and throwing this error,
Its seems to me that I missed some step to configure my library in the project,
can somebody suggest a steps or guideline how to consume angular library in project angular-cli?

**Comment by chrillewoodz** [comment]
To all these people telling u to "remove the node modules folder in the lib", that will completely remove the ability to develop the lib and test what you're doing at the same time.. Since you need the lib's deps installed in order to build etc.

**Comment by btgoodwin** [comment]
What I ultimately did for my CLI app build was to start adding additional scripts for `build` and `start` when doing development against my libraries:

```json
"build:dev": "npm run build -- --preserve-symlinks",
"start:dev": "npm run start -- --preserve-symlinks"
```

That gets around the problem without having to do exotic things like build my `dist` directory outside of the library's source tree (and then link it from there).  

I'm using the [generator-angular2-library](https://github.com/jvandemo/generator-angular2-library) to create my libraries, which works great if anyone is interested (I'm not affiliated with their work at all other than my PR to add auto version migration to the `dist/package.json` on `npm version ...`).

**Comment by Basel78** [comment]
Actually, I have fixed the issues and was able to run my library in new project.
part of the fix was two things:
1- remove the line: import "reflect-metadata";  from common.ts file Common library.
2- using export class with static function like ForRoot() to return ModuleWithProviders for NgModule instead of just exporting normal class,
not sure why the first point caused that error, and it took me two days to isolate my code line by line to find once that caused that issue and fix it,

**Comment by magemello** [comment]
I work with my team on a complex project developed in Angular : https://github.com/Alfresco/alfresco-ng2-components.

our project is structured in this way:
- angular-cli-demo-app // demo app which we use to test our components
- components
    - alfresco-angular-component 1 // component library based on Angular
    - alfresco-angular-component 2
    - alfresco-angular-component 3 

Inside the "angular-cli-demo-app" we refer all our alfresco-angular-component using the  tsconfig.json path: 
```
"compilerOptions" :{... 
 "path"{}
}
```
Until we didn't use the path reference strategy everything worked fine.

**The way to fix this problem for us was to delete all the @angular folder inside the node_modules in our components and all the shared folder**, because even if we referred it in the path tsconfig.json for some reason our component where still following the npm closer parent node_modules folder rule to get the @angular library.
This of course resulted in having multiple @angular running in the project with multiple @NgModule wichs cause the error: 
`Please add a @NgModule annotation.:`

_I hope this explanation save some debug time to the others, and also our project offers some reference to who wants to build a project + components library._

**Comment by m0v0nage** [comment]
Is there an issue open that has the aim to fix this issue, in this project or another? @magemello has a good workaround, but I find myself having to delete my node_modules often when making changes/adding new dependencies in my components library. Why aren't the paths in `tsconfig.json` being respected?

**Comment by linktothapast** [comment]
I've tried everything mentioned in this issue to no avail. Using the latest Angular packages (`5.0.1`), no matter what I try, I keep getting the same error:
```
Unexpected value 'MyCustomModule' imported by the module 'AppModule'. Please add a @NgModule annotation.
```
- There is no internal node_modules inside `node_modules/my-custom-module`
- All `@angular/*` dependencies are specified as `peerDependencies`
- Tried different configuration settings in the custom library's `tsconfig.json`. Tried `skipTemplateCodegen` both true and false (because of @robwormald comment above)

This was all working fine before.

**Comment by IevhenIkonnykov** [comment]
@linktothapast  Unfortunately this is a very sneaky error that can be caused by different reasons. The last time I experienced this when I was adding a module to the root module and it had BrowserModule and BrowserAnimationsModule imported in it and this was my mistake. At least be sure that two above mentioned modules imported only once and only in the root module, otherwise you can have errors that are not very descriptive. Use CommonModule in your other modules instead of BrowserModule. In my case I had the same error as you and it was about the module that was totally fine. I hope it might be helpful. Also please check your imports, declarations, exports that everything on its place. I mean there is not a case that you declared a module in declarations or imported service etc. These kinds of mistake might cause weird errors as well.

**Comment by bentedder** [comment]
Similar to @linktothapast I have tried everything I can think of. I've created 2 simple angular 5 projects, one to package up, and one to consume. I use `npm link` to connect them, and I've put the files in a Plunkr for the packaged file to see if anybody knows what I'm doing wrong. Can't think of any other way but rolling back and waiting a few more weeks until smarter people than me run into the problem and post a fix :). The projects themselves build and run fine, it's only when exporting for library usage.

https://plnkr.co/edit/bOzJh5ncDSBKGoHggfDI?p=info

**Comment by omouse** [comment]
@benelliott yep, I've been waiting for a solution for months now and this kind of stuff is a blocker to working with Angular. While I love promoting Angular for larger web dev teams because at its core it *is* a good framework, all these infrastructure issues and all the work that the Angular team has taken on is leading to lots of time wasted debugging issues like this.

**Comment by mmastrangelo** [comment]
I am having this problem as well. We will be stuck on version 1.2.7 of the CLI until a solution is available. Note that this issue is particularly challenging for anyone on NPM 5 as it links projects by default that are specified via "file:" in package.json. The only way to make it work is to remove node_modules from my linked modules... which is not viable within our dev workflow.

**Comment by dohpaz42** [comment]
This is a closed ticket; not sure the devs look at replies to closed tickets. Maybe time to open a new ticket?

Also, I wanted to point out that, at least for me, I found and mirrored setting up my modules as demonstrated in the [ngx-rocket/starter-kit](https://github.com/ngx-rocket/starter-kit/tree/master/src/app) and it has worked well ever since. The only gotcha is when my IDE decides to use the shortcut notation (e.g., @shared/whatever) inside of a module when it shouldn't.

**Comment by lazarljubenovic** [comment]
Same issue here. Had a library written in `4.4.4`. After updating to `5.0.x`, the only change I made was tweak the build process to use the new function signature of `ngc`, and then this error started popping up. I've check out the metadata file that I'm generating and it looks pretty fine to me:

```
{
  "__symbolic": "module",
  "version": 4,
  "metadata": {
    "YahteeDatePickerModule": {
      "__symbolic": "class",
      "decorators": [
        {
          "__symbolic": "call",
          "expression": {
            "__symbolic": "reference",
            "module": "@angular/core",
            "name": "NgModule"
          },
          "arguments": [
            {
              "imports": [
                {
                  "__symbolic": "reference",
                  "module": "@angular/common",
                  "name": "CommonModule"
                }
              ],
```

I am consuming the library via npm repo (no links involved, like a few comments above pointed out). It was all working fine when the lib was built using 4.x and consumed via 5.x. Now when the lib is built wit 5.x and used via 5.x, it breaks runtime with the following error.

```
ERROR Error: Uncaught (in promise): Error: Unexpected value 'YahteeDatePickerModule' imported by the module 'DatePickerModule'. Please add a @NgModule annotation.
Error: Unexpected value 'YahteeDatePickerModule' imported by the module 'DatePickerModule'. Please add a @NgModule annotation.
```

The consuming project is using CLI 1.5.

**Comment by rudzikdawid** [comment]
i have similar problem with `@angular/cli 1.4.9` here https://github.com/swiety85/angular2gridster/tree/137-demo-src-app-gridster

dir structure:
```
angular2gridster
├─ demo //angular-cli app
│  └─ src
└─ src //library source
```

reproduce steps:
```
cd angular2gridster
npm link
cd demo
npm link angular2gridster 
```
![falubas](https://i.imgur.com/NaYBcxe.png)
after that in `demo/node_modules` we have linked `angular2gridster` package but with library's `node_modules` which unfortunately contain `@angular/core` 

start demo `ng serve --preserve-symlinks` affects error:
```
Uncaught Error: Unexpected value 'GridsterModule'
imported by the module 'AppModule'. Please add a @NgModule annotation.
```

## messy workaround
removing `demo/node_modules/angular2gridster/node_modules/@angular/core`
solves the problem temporary but after that we can't rebuild library.


path mapping inside `demo/src/tsconfig.app.json` also doesn't help:
```
"baseUrl": ".",
"paths": {
   "@angular/core": ["../node_modules/@angular/core"]
 }
```


library dependecies:
```
"peerDependencies": {
    "@angular/common": "4.4.6",
    "@angular/core": "4.4.6"
  },
  "devDependencies": {
    "@angular/common": "4.4.6",
    "@angular/compiler": "4.4.6",
    "@angular/compiler-cli": "4.4.6",
    "@angular/core": "4.4.6",
    "@angular/language-service": "4.4.6",
    "@types/node": "~6.0.60",
    "codelyzer": "^3.0.1",
    "reflect-metadata": "~0.1.9",
    "rxjs": "^5.5.2",
    "ts-loader": "^2.0.1",
    "ts-node": "~3.0.4",
    "tslint": "~5.3.2",
    "typescript": "2.4.2",
    "zone.js": "^0.8.4"
  }
```


demo dependencies:
```
"dependencies": {
    "@angular/animations": "4.4.6",
    "@angular/common": "4.4.6",
    "@angular/compiler": "4.4.6",
    "@angular/core": "4.4.6",
    "@angular/forms": "4.4.6",
    "@angular/http": "4.4.6",
    "@angular/platform-browser": "4.4.6",
    "@angular/platform-browser-dynamic": "4.4.6",
    "@angular/router": "4.4.6",
    "core-js": "^2.5.1",
    "rxjs": "^5.5.2",
    "zone.js": "^0.8.4"
  },
  "devDependencies": {
    "@angular/cli": "1.4.9",
    "@angular/compiler-cli": "4.4.6",
    "@angular/language-service": "4.4.6",
    "@types/jasmine": "~2.5.53",
    "@types/jasminewd2": "~2.0.2",
    "@types/node": "~6.0.60",
    "codelyzer": "~3.0.1",
    "gh-pages": "^1.1.0",
    "jasmine-core": "~2.6.2",
    "jasmine-spec-reporter": "~4.1.0",
    "karma": "~1.7.0",
    "karma-chrome-launcher": "~2.1.1",
    "karma-cli": "~1.0.1",
    "karma-coverage-istanbul-reporter": "^1.2.1",
    "karma-jasmine": "~1.1.0",
    "karma-jasmine-html-reporter": "^0.2.2",
    "protractor": "~5.1.2",
    "ts-node": "~3.0.4",
    "tslint": "~5.3.2",
    "typescript": "2.4.2"
  }
```

any idea?

**Comment by OnlyBelter** [comment]
I also met this problem: same code works on Windows, but not works on Linux. 


This is my version infomation:
### windows
```
Angular CLI: 1.5.0
Node: 8.1.4
OS: win32 x64
Angular: 4.3.6
... common, compiler, compiler-cli, core, forms, http
... language-service, platform-browser, platform-browser-dynamic
... router, tsc-wrapped

@angular/animations: 4.4.6
@angular/cli: 1.5.0
@angular-devkit/build-optimizer: 0.0.32
@angular-devkit/core: 0.0.20
@angular-devkit/schematics: 0.0.35
@ngtools/json-schema: 1.1.0
@schematics/angular: 0.1.1
typescript: 2.3.4
webpack: 3.8.1
```

### linux
```
Angular CLI: 1.5.0
Node: 6.11.3
OS: linux x64
Angular: 4.4.6
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router, tsc-wrapped

@angular/cli: 1.5.0
@angular-devkit/build-optimizer: 0.0.32
@angular-devkit/core: 0.0.20
@angular-devkit/schematics: 0.0.35
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.8.0
@schematics/angular: 0.1.1
typescript: 2.3.4
webpack: 3.8.1
```

Then I updated the `nodejs` on linux from `6.11.3` to `8.9.3`, deleted all dependencies then used `npm install` to re-install all dependencies. It works now.

**Comment by bDino** [comment]
Had the same problem - turned out the ng-pdf-make project had a subdirectory of node_modules with its own angular version which caused the bug. After deleting the local node_modules it worked!

**Comment by srinivas444** [comment]
I also faced the same issue but resolved by updating the tsconfig.app.json with following code:

{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../out-tsc/app",
    "module": "es2015",
    "baseUrl": "",
    "types": [],
    "typeRoots": [
"node_modules/@types"
],
"lib": [
"es2017",
"dom"
],
    "paths": {
      "@angular/*": [
      "../node_modules/@angular/*"
      ]
    }
  },
  "exclude": [
    "test.ts",
    "**/*.spec.ts"
  ]
}

Its worked for me try it from your end.

**Comment by raviprajna** [comment]
I am also facing same issue 
--------------------------------------

Uncaught Error: Unexpected value 'Http' imported by the module 'AppModule'. Please add a @NgModule annotation.
    at syntaxError (compiler.js:215)
    at compiler.js:10468
    at Array.forEach (<anonymous>)
    at CompileMetadataResolver.push../node_modules/@angular/compiler/fesm5/compiler.js.CompileMetadataResolver.getNgModuleMetadata (compiler.js:10443)
    at JitCompiler.push../node_modules/@angular/compiler/fesm5/compiler.js.JitCompiler._loadModules (compiler.js:22567)
    at JitCompiler.push../node_modules/@angular/compiler/fesm5/compiler.js.JitCompiler._compileModuleAndComponents (compiler.js:22548)
    at JitCompiler.push../node_modules/@angular/compiler/fesm5/compiler.js.JitCompiler.compileModuleAsync (compiler.js:22508)
    at CompilerImpl.push../node_modules/@angular/platform-browser-dynamic/fesm5/platform-browser-dynamic.js.CompilerImpl.compileModuleAsync (platform-browser-dynamic.js:143)
    at PlatformRef.push../node_modules/@angular/core/fesm5/core.js.PlatformRef.bootstrapModule (core.js:4790)
    at Object../src/main.ts (main.ts:11)

**Comment by BenceUszkai** [comment]
try this: https://medium.com/@geeksamu/using-angular-material-with-angular-6-862a1c3d04b

**Comment by Jayawardhana** [comment]
i used flash-message new version then i have this error in browser

[WDS] Disconnected!
vendor.bundle.js:67956:10
Error: Unexpected module 'FlashMessageModule' declared by the module 'AppModule'
zone.js:992

This is app.module.ts

// Required Fields
    if(!this.validateService.validateRegister(user)){
      this.flashMessage.warning('Please fill in all fields', {cssClass:'alert-danger', timeout:3000});
      return false;
    }

    // Validate Email
    if(!this.validateService.validateEmail(user.email)){
      this.flashMessage.warning('Please use a valid email', {cssClass:'alert-danger', timeout:3000});
      return false;
    }

**Comment by lazarljubenovic** [comment]
@Jayawardhana This is not a support group. Please post your question on Stack Overflow. And it does make sense, you do not declare a module in a module, you import it.

**Comment by rsprince** [comment]
Unhelpful and negative.

**Comment by lazarljubenovic** [comment]
It's not negative, it's a fact. Hijacking an issue with a "fix this code please" only creates noise for everyone. I've not only pointed the user in the right direction (StackOverflow), I've actually given a strong hint towards the solution:

> you do not declare a module in a module, you import it

So there's nothing negative about it and there's for sure a few things helpful there.

**Comment by vilbara** [comment]
Got the same error with angular v7 with an in-project library created with angular-cli. The cause in my case was that the components and modules were exported in _public_api.ts_ via intermediate _index.ts_ files. Removing _index.ts_ files and exporting everything directly in _public_api.ts_ solved this.

Pretty sure there is a good technical explanation for this, but from the perspective of a simple "user" this belongs to the category of "insane"!

**Comment by nejsimon** [comment]
> Got the same error with angular v7 with an in-project library created with angular-cli. The cause in my case was that the components and modules were exported in _public_api.ts_ via intermediate _index.ts_ files. Removing _index.ts_ files and exporting everything directly in _public_api.ts_ solved this.

Thanks! This solved the problem for me as well. The error message you get is remarkably unhelpful.

**Comment by yanqiw** [comment]
> Got the same error with angular v7 with an in-project library created with angular-cli. The cause in my case was that the components and modules were exported in _public_api.ts_ via intermediate _index.ts_ files. Removing _index.ts_ files and exporting everything directly in _public_api.ts_ solved this.
> 
> Pretty sure there is a good technical explanation for this, but from the perspective of a simple "user" this belongs to the category of "insane"!

this one save me.

**Comment by kimamula** [comment]
> Got the same error with angular v7 with an in-project library created with angular-cli. The cause in my case was that the components and modules were exported in public_api.ts via intermediate index.ts files. Removing index.ts files and exporting everything directly in public_api.ts solved this.

Thanks!
Specifying `index.ts` explicitly in `public_api.ts` (i.e., `export * from './some-module/index'`) also works.
